ext/intl/dateformat/datepatterngenerator_methods.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Mel Dafert (mel@dafert.at)                                  |
+   +----------------------------------------------------------------------+
+*/
+
+#include "../intl_cppshims.h"
+
+#include <unicode/dtptngen.h>
+
+#include "../intl_convertcpp.h"
+
+extern "C" {
+#include "php_intl.h"
+#define USE_DATETIMEPATTERNGENERATOR_POINTER 1
+#include "datepatterngenerator_class.h"
+#include <zend_exceptions.h>
+#include <assert.h>
+}
+
+using icu::DateTimePatternGenerator;
+using icu::Locale;
+using icu::StringPiece;
+
+static zend_result dtpg_ctor(INTERNAL_FUNCTION_PARAMETERS)
+{
+	zval		*object;
+	char		*locale_str;
+	size_t		locale_len	= 0;
+	Locale		locale;
+	IntlDatePatternGenerator_object* dtpgo;
+
+	intl_error_reset(NULL);
+	object = return_value;
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_STRING_OR_NULL(locale_str, locale_len)
+	ZEND_PARSE_PARAMETERS_END_EX(return FAILURE);
+
+	DTPATTERNGEN_METHOD_FETCH_OBJECT_NO_CHECK;
+
+	if (dtpgo->dtpg != NULL) {
+		intl_errors_set(DTPATTERNGEN_ERROR_P(dtpgo), U_ILLEGAL_ARGUMENT_ERROR, "Cannot call constructor twice", 0);
+		return FAILURE;
+	}
+
+	INTL_CHECK_LOCALE_LEN_OR_FAILURE(locale_len);
+	if (locale_len == 0) {
+		locale_str = (char *) intl_locale_get_default();
+	}
+	locale = Locale::createFromName(locale_str);
+
+	dtpgo->dtpg = DateTimePatternGenerator::createInstance(
+		locale,
+		DTPATTERNGEN_ERROR_CODE(dtpgo));
+
+	if (U_FAILURE(DTPATTERNGEN_ERROR_CODE(dtpgo))) {
+		intl_error_set(NULL, DTPATTERNGEN_ERROR_CODE(dtpgo),
+				"Error creating DateTimePatternGenerator",
+				0);
+		goto error;
```

Comment:
This goto error seems pretty pointless, seeing how it goes to the next line ^^

Reply:
Though can you just return FAILURE here and unconditionally return SUCCESS below?